libraries/chain/controller.cpp

```diff
@@ -3961,7 +3964,7 @@ struct controller_impl {
 
       const bool skip_validate_signee = false;
       auto bsp = std::make_shared<BS>(
-            prev,
+            *prev,
             b,
```

Comment:
Prior to calling this constructor, we need to at least make sure that the QC claim in the block header of block `b` does not go backwards relative to the QC claim in the previous block (`prev`). 

This is needed to satisfy the precondition `this->latest_qc_claim() <= most_recent_ancestor_with_qc` assumed in `finality_core::next` and `finality_core::next_metadata` which is (indirectly) called from this constructor.

Previously that condition was checked in `verify_qc_claim`. If it was not satisfied, it threw an exception to reject the block and exited earlier before getting to this part of the code.

Now that it is posted to another thread and the result (throwing an exception or not) is only guaranteed by the point `verify_future.get()` is called below, we cannot deterministically guarantee that the precondition will be satisfied by the time this constructor is called.

At the very least we would need check the QC claim earlier in this function (outside of the future) to guarantee the precondition, and the remainder of the checks in `verify_qc_claim` can be done later as long as it is prior to calling `integrate_received_qc_to_block`.